Improve docs structure

mauriciovasquezbernal opened the pull request:
I'm opening this PR as a Request For Comments to understand if it makes sense to continue with this approach to improve documentation or not.

The problems I'm trying to solve are:
- Current documentation structure is quite plain, an index document with a lot of links to different modules.
- There are not examples in [the current online documentation](https://opentelemetry-python.readthedocs.io/en/latest/).
- It's difficult to keep aligned the examples with the latest released version, i.e, examples in main readme are not synchronized with packages versions on PyPI.

This PR:
- Creates a tree structure for documentation, it allows to organize it better from a developer point of view and also the rendered documentation should be easier to navigate.
- Moves partially the main readme to be included in the online docs, the main readme will be updated to have a link to avoid duplicated content)
- Moves the examples folder to the docs, so they can be accessed through the online documentation. 
- Creates a new pair of "macros" to create links to specific versions. `scm_web` & `scm_raw_web`.

The result documentation can be accessed at https://opentelemetry-python-mauricio.readthedocs.io/en/improve-docsv1/index.html, the only updated example is available at https://opentelemetry-python-mauricio.readthedocs.io/en/improve-docsv1/examples/basic_tracer/README.html. (notice that it includes a link to the example source that is dynamic according to the version).

Please let me know your feedback, I'll happy to continue working on this if you think the result it worthy.

toumorokoshi commented on docs/conf.py:
```diff
@@ -106,3 +108,22 @@
+if branch == None or branch == 'latest':
+    branch = 'master'
+
+scm_raw_web = 'https://raw.githubusercontent.com/mauriciovasquezbernal/opentelemetry-python/' + branch
```
looks like these should be pointed at the main telemetry repo.

mauriciovasquezbernal replied:
Will update just before merging, otherwise I cannot see the results on my testing readthedocs instance.

toumorokoshi commented on docs/examples/basic_tracer/README.rst:
```diff
@@ -0,0 +1,83 @@
+
+The output will be displayed at the console
+
+.. code-block:: bash
```
why is this block bash, while the above is sh?

toumorokoshi commented on docs/examples/basic_tracer/README.rst:
```diff
@@ -0,0 +1,83 @@
+
+
+Jaeger UI
+*********
```
is this section necessary? seems to me like we should just assume that people know how to use jaeger, and save ourselves docs that we'll have to update later.

mauriciovasquezbernal replied:
I removed it, let's keep this simple.

toumorokoshi commented on docs/examples/index.rst:
```diff
@@ -0,0 +1,6 @@
+    :maxdepth: 1
+    :caption: examples
+
+    basic_tracer/README
```
should we use a broader toctree expression to captures these? IIRC we can use glob. Would save someone having to add to this file every time an example is added.

toumorokoshi commented on docs/index.rst:
```diff
@@ -1,47 +1,98 @@
-.. toctree::
-   :maxdepth: 1
-   :caption: OpenTelemetry API:
+TODO: intro with link to opentelemetry.io
```
can we remove TODOs, or use sphinx comments? 

I'd say the omission of data isn't a big detail, we can always file subsequent PRs to fill that in.

toumorokoshi commented on docs/index.rst:
```diff
@@ -1,47 +1,98 @@
-   :caption: OpenTelemetry API:
+TODO: intro with link to opentelemetry.io
+
+Installation
```
FWIW, I wrote an index page that's similar, but has some wording changed to better reflect the fact that this is in a doc vs the repo itself. I'll post here and you can pull stuff in, or I can come in after and add parts I feel are valuable.

```
.. OpenTelemetry documentation master file, created by
   sphinx-quickstart on Mon Jun  3 22:48:38 2019.
   You can adapt this file completely to your liking, but it should at least
   contain the root `toctree` directive.

OpenTelemetry-Python
====================

The Python `OpenTelemetry <https://opentelemetry.io/>`_ client.

.. image:: https://img.shields.io/gitter/room/opentelemetry/opentelemetry-python
   :target: https://gitter.im/open-telemetry/opentelemetry-python
   :alt: Gitter Chat


This documentation describes the opentelemetry-api, opentelemetry-sdk,
and several integration packages.

**Please note** that this library is currently in alpha, and shouldn't be
used in production environments.

Installation
------------

The API and SDK packages are available on PyPI, and can installed via pip:

.. code-block:: bash

    pip install opentelemetry-api
    pip install opentelemetry-sdk

In addition, there are several extension packages which can be installed separately as::

    pip install opentelemetry-ext-{integration}

The extension packages can be found in
`ext/ directory of the repository <https://github.com/open-telemetry/opentelemetry-python/tree/master/ext>`_.

In addition, third party exporters are available:

* `Azure Monitor <https://github.com/microsoft/opentelemetry-exporters-python/tree/master/azure_monitor>`_

Installing Cutting Edge Packages
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

While the project is pre-1.0, there may be significant functionality that
has not yet been released to PyPI. In that situation, you may want to 
install the packages directly from the repo. This can be done by cloning the
repositry and doing an `editable
install <https://pip.pypa.io/en/stable/reference/pip_install/#editable-installs>`_:

.. code-block:: bash
    git clone https://github.com/open-telemetry/opentelemetry-python.git
    cd opentelemetry-python
    pip install -e ./opentelemetry-api
    pip install -e ./opentelemetry-sdk
    pip install -e ./ext/opentelemetry-ext-{integration}


Quick Start
-----------

opentelemetry can be used to emit distributed traces and metrics from your application. 
The following are examples using the API and SDK:

Tracing
~~~~~~~

.. code-block:: python

    from opentelemetry import trace
    from opentelemetry.sdk.trace import TracerProvider
    from opentelemetry.sdk.trace.export import ConsoleSpanExporter
    from opentelemetry.sdk.trace.export import SimpleExportSpanProcessor

    trace.set_preferred_tracer_provider_implementation(lambda T: TracerProvider())
    trace.tracer_provider().add_span_processor(
        SimpleExportSpanProcessor(ConsoleSpanExporter())
    )
    tracer = trace.get_tracer(__name__)
    with tracer.start_as_current_span('foo'):
        with tracer.start_as_current_span('bar'):
            with tracer.start_as_current_span('baz'):
                print("Hello world from OpenTelemetry Python!")

Metrics
~~~~~~~

.. code-block:: python

    from opentelemetry import metrics
    from opentelemetry.sdk.metrics import Counter, MeterProvider
    from opentelemetry.sdk.metrics.export import ConsoleMetricsExporter
    from opentelemetry.sdk.metrics.export.controller import PushController

    metrics.set_preferred_meter_provider_implementation(lambda _: MeterProvider())
    meter = metrics.get_meter(__name__)
    exporter = ConsoleMetricsExporter()
    controller = PushController(meter, exporter, 5)

    counter = meter.create_metric(
        "available memory",
        "available memory",
        "bytes",
        int,
        Counter,
        ("environment",),
    )

    label_values = ("staging",)
    counter_handle = counter.get_handle(label_values)
    counter_handle.add(100)

More examples are available in the `examples folder <https://github.com/open-telemetry/opentelemetry-python/tree/master/examples>`_.


.. toctree::
   :maxdepth: 1
   :caption: OpenTelemetry API:

   opentelemetry.context
   opentelemetry.metrics
   opentelemetry.trace
   opentelemetry.util.loader

.. toctree::
    :maxdepth: 1
    :caption: OpenTelemetry SDK:

    opentelemetry.sdk.context
    opentelemetry.sdk.metrics
    opentelemetry.sdk.trace


.. toctree::
    :maxdepth: 1
    :caption: OpenTelemetry Integrations:

    opentelemetry.ext.flask
    opentelemetry.ext.http_requests
    opentelemetry.ext.jaeger
    opentelemetry.ext.opentracing_shim
    opentelemetry.ext.pymongo
    opentelemetry.ext.wsgi

Indices and tables
==================

* :ref:`genindex`
* :ref:`modindex`
* :ref:`search`

```

toumorokoshi commented on docs/sdk/sdk.rst:
```diff
@@ -0,0 +1,11 @@
+OpenTelemetry Python SDK
+========================
+
+TODO: what is the SDK
```
again I'd vote to comment the TODO.

toumorokoshi commented on opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py:
```diff
@@ -136,9 +136,9 @@ class Span(trace_api.Span):
-    empty_attributes = BoundedDict(MAX_NUM_ATTRIBUTES)
-    empty_events = BoundedList(MAX_NUM_EVENTS)
-    empty_links = BoundedList(MAX_NUM_LINKS)
+    _empty_attributes = BoundedDict(MAX_NUM_ATTRIBUTES)
```
was this to help the autodocumentation work better?

mauriciovasquezbernal replied:
Yes, it was just a detail I found. Without the underscore those identifiers are added to the documentation, I don't see any reason to have them there.

mauriciovasquezbernal commented:
I updated the PR to fix the tests (I had to disable the warning as errors in tox docs)

I also added documentation for external packages. The readme of each package is included + the autodoc documentation. I think it's just a temporal solution,  I think the best approach is:
- Document the package inside the code, like opentracing shim https://github.com/open-telemetry/opentelemetry-python/blob/9ed98eb9320b9064e43c3b43ee7c4990eec3657a/ext/opentelemetry-
ext-opentracing-shim/src/opentelemetry/ext/opentracing_shim/__init__.py#L16
- Autodoc will take care of all documentation because of ^^^.
- Use a minimal readme with a link to readthedocs (Will be shown in PyPI).

I don't want to add more things to this PR right now, we can improve this in follow up PRs.

c24t commented on docs/index.rst:
```diff
@@ -1,47 +1,104 @@
+Tracing
+~~~~~~~
+
+.. literalinclude:: trace_example.py
```
Good call, transclusion here really helps to clean up the docs.

c24t commented on docs/examples/opentracing/README.rst:
```diff
@@ -0,0 +1,97 @@
+OpenTracing Bridge Example
```
```suggestion
OpenTracing Shim Example
```

To match the package name, and because people may interpret "shim" and "bridge" to have different meanings.

c24t commented on docs/index.rst:
```diff
@@ -1,47 +1,104 @@
-   :maxdepth: 1
-   :caption: OpenTelemetry API:
+This documentation describes the :doc:`opentelemetry-api <api/api>`,
+:doc:`opentelemetry-sdk <sdk/sdk>`, and several `integration packages <#integrations>`_
```
```suggestion
:doc:`opentelemetry-sdk <sdk/sdk>`, and several `integration packages <#integrations>`_.
```

c24t commented on docs/index.rst:
```diff
@@ -1,47 +1,104 @@
+Quick Start
+-----------
+
+opentelemetry can be used to emit distributed traces and metrics from your application.
```
```suggestion
OpenTelemetry can be used to emit distributed traces and metrics from your application.
```

Or format as code.

c24t commented on docs/conf.py:
```diff
@@ -106,3 +108,27 @@
 html_static_path = []
+
+# Support external links to specific versions of the files in the Github repo
+branch = os.environ.get("READTHEDOCS_VERSION")
```
Nice!

mauriciovasquezbernal commented on docs/conf.py:
```diff
@@ -106,3 +108,27 @@
+
+REPO = "open-telemetry/opentelemetry-python/"
+scm_raw_web = "https://raw.githubusercontent.com/" + REPO + branch
+scm_web = "https://github.com/" + REPO + "blob/" + branch
```
Yes, it's a sphinx feature. It'll also not render correctly on Github, for instance: https://github.com/mauriciovasquezbernal/opentelemetry-python/tree/improve-docsv1/docs/examples/basic_tracer. I don't have that many ideas to solve that, I think there is not a perfect solution.

toumorokoshi commented:
@mauriciovasquezbernal let me know when this is ready to merge. Thanks for all the improvements here! a huge win.

mauriciovasquezbernal commented:
@toumorokoshi it's ready to go. It's far from perfect but we can work to improve it later on.
